src/Workspaces/Core/Portable/CodeActions/CodeAction_Cleanup.cs

```diff
@@ -0,0 +1,180 @@
+{
+    /// <summary>
+    /// We do cleanup in N serialized passes.  This allows us to process all documents in parallel, while only forking
+    /// the solution N times *total* (instead of N times *per* document).
+    /// </summary>
+    private static readonly ImmutableArray<Func<Document, CodeCleanupOptions, CancellationToken, Task<Document>>> s_cleanupPasses =
+    [
+        // First, ensure that everything is formatted as the feature asked for.  We want to do this prior to doing
+        // semantic cleanup as the semantic cleanup passes may end up making changes that end up dropping some of
+        // the formatting/elastic annotations that the feature wanted respected.
+        static (document, options, cancellationToken) => CleanupSyntaxAsync(document, options, cancellationToken),
+        // Then add all missing imports to all changed documents.
+        static (document, options, cancellationToken) => ImportAdder.AddImportsFromSymbolAnnotationAsync(document, Simplifier.AddImportsAnnotation, options.AddImportOptions, cancellationToken),
+        // Then simplify any expanded constructs.
+        static (document, options, cancellationToken) => Simplifier.ReduceAsync(document, Simplifier.Annotation, options.SimplifierOptions, cancellationToken),
+        // The do any necessary case correction for VB files.
+        static (document, options, cancellationToken) => CaseCorrector.CaseCorrectAsync(document, CaseCorrector.Annotation, cancellationToken),
+        // Finally, after doing the semantic cleanup, do another syntax cleanup pass to ensure that the tree is in a
+        // good state. The semantic cleanup passes may have introduced new nodes with elastic trivia that have to be
+        // cleaned.
+        static (document, options, cancellationToken) => CleanupSyntaxAsync(document, options, cancellationToken),
+    ];
+
+    private static async Task<Document> CleanupSyntaxAsync(Document document, CodeCleanupOptions options, CancellationToken cancellationToken)
+    {
+        Contract.ThrowIfFalse(document.SupportsSyntaxTree);
+
+        // format any node with explicit formatter annotation
+        var document1 = await Formatter.FormatAsync(document, Formatter.Annotation, options.FormattingOptions, cancellationToken).ConfigureAwait(false);
+
+        // format any elastic whitespace
+        var document2 = await Formatter.FormatAsync(document1, SyntaxAnnotation.ElasticAnnotation, options.FormattingOptions, cancellationToken).ConfigureAwait(false);
+        return document2;
+    }
+
+    internal static ImmutableArray<DocumentId> GetAllChangedOrAddedDocumentIds(
+        Solution originalSolution,
+        Solution changedSolution)
+    {
+        var solutionChanges = changedSolution.GetChanges(originalSolution);
+        var documentIds = solutionChanges
+            .GetProjectChanges()
+            .SelectMany(p => p.GetChangedDocuments(onlyGetDocumentsWithTextChanges: true).Concat(p.GetAddedDocuments()))
```

Comment:
>GetChangedDocuments [](http://example.com/codeflow?start=31&length=19)

We don't need to worry about additional or analyzerconfig docs, right?

Reply:
nope. those are never cleaned up.  'cleaning' is a specific C#/VB concept where we do final passes, applying formatting, and going and simplifying code that a code fixer annotated in the syntax trees.